### PR TITLE
Remove HasParent (involves moving glyph display_defaults into defaults)

### PR DESCRIPTION
**Pull request by canavandl**
issues: #3651 

Two things:
* ~~Remove AbstractRendering cruft (delete bokehjs/src/coffee/ar)~~ moved into distinct branch
* Remove HasParent (have subclasses inherit from HasProperties and move Glyph display_defaults into defaults)

### CODE REVIEW

**Comment by bryevdv** [comment]
I think a better PR breakdown might be:

* remove AR cruft

* remove `HasParent` (includes display defaults changes)

**Comment by canavandl** [comment]
Yup - you're too quick for me.

**Comment by bryevdv** [comment]
Well I was actually a bit vague, but I meant the AR stuff could go in a separate PR.

**Comment by bryevdv** [comment]
Would make a good excuse to try out `git cherry-pick` if you haven't used it before :)

**Comment by canavandl** [comment]
I could also be less lazy and make it 2 PRs (better for my GH statz)

**Comment by bryevdv** [comment]
I'm communicating badly today :) two PRs is what I meant (just that `git cherry-pick` will let you make two new PRs out of the bones of this one very easily)

**Comment by canavandl** [comment]
Haha - no I'm just having reading comprehension problems.

**Comment by bryevdv** [comment]
Anything else to do on this PR?

**Comment by canavandl** [comment]
Nope - ready for review

**Comment by bryevdv** [comment]
LGTM any comments @mattpap  or @havocp ?

File: bokehjs/test/test_common/test_defaults.coffee [hunk 1]
```diff
@@ -136,39 +136,11 @@ describe "Defaults", ->
...
-            if not (_.isObject(orig) and ('field' of orig or 'value' of orig))
-              # convert to 'spec dict' format
-              display[k] = { 'value' : display[k] }
-
-        attrs = _.extend({}, display, attrs)
-
-      # Merge in "factory" attrs, which don't go in attributes
-      # in the actual code, but for our current purpose
-      # let's pretend they do because they do "in effect"
-      # and we probably want to put them there eventually
```
Comment (havocp) [inline bokehjs/test/test_common/test_defaults.coffee]: the above comment is about code you're keeping I think?

File: bokehjs/test/test_common/test_defaults.coffee [hunk 2]
```diff
@@ -136,39 +136,11 @@ describe "Defaults", ->
...
+
       for prop_kind, func of properties.factories
-        if prop_kind == 'visuals' # visuals is the display_defaults() case above
+        # the 'visuals' property is used to set glyph line/fill/text defaults
+        # and is tested in the check_matching_defaults() test case above
```
Comment (havocp) [inline bokehjs/test/test_common/test_defaults.coffee]: this comment doesn't make sense to me - are the visual props DataSpecs? if so then we probably shouldn't skip them here anymore. if not the comment could be "visuals aren't data specs so don't need special processing here"

**Comment by havocp** [comment]
glad to see this code go, nice

File: bokehjs/test/test_common/test_defaults.coffee (hunk 2 above)
Comment (bryevdv) [inline bokehjs/test/test_common/test_defaults.coffee]: The visual properties comprise line, fill, and text props. Which then expand into collections of individual properties, some of which are data specs. c.f. https://github.com/bokeh/bokeh/blob/master/bokeh/core/property_mixins.py#L39 for instance, `line_width` is a spec (i.e. vectorizable) but `line_dash` is not

**Comment by damianavila** [comment]
@canavandl just a tiny comment... you are putting `issues: fixes #3651` at the top of this and others... in act #3651 is a meta task issue... if you put this reference again... when this will merge the issue will be closed again automatically...
I suggest to not use the `fixes` when you are dealing with meta task issues... just reference the issue and make this one an standalone issues-PR... I just labeled this one as is... thanks!!

File: bokehjs/test/test_common/test_defaults.coffee (hunk 2 above)
Comment (bryevdv) [inline bokehjs/test/test_common/test_defaults.coffee]: I'm actually not sure offhand but I'd like to have a dedicated PR to bring clean up JS properties and bring the implementation more in line witth the python side (to the extend possible w/o multiple inheritance mixing). Let's figure this out there.

**Pull request merged**